src/irr/core/CFixedCapacityDoublyLinkedList.h

```diff
@@ -0,0 +1,161 @@
+				return *this;
+			}
+		};
+
+
+		template<typename Value>
+		class DoublyLinkedList
+		{
+
+			void* reservedSpace;
+			PoolAddressAllocator<uint32_t> alloc;
+			uint32_t m_back;
+			uint32_t m_begin;
+			Snode<Value>* m_array;
+			uint32_t cap;
+
+		public:
+			_IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_iterator = 0xdeadbeefu;
+
+			//get node at iterator
+			inline Snode<Value>* get(uint32_t address)
+			{
+				return (m_array+address);
+			}
+
+			//remove the last element in the list
+			inline void popBack()
+			{
+				if (m_back == invalid_iterator)	return;
+				auto backNode = getBack();
+				if (backNode->prev != invalid_iterator)
+					get(backNode->prev)->next = invalid_iterator;
+				uint32_t temp = m_back;
+				m_back = backNode->prev;
+				alloc.free_addr(temp, 1u);
+			}
+
+			//add new item to the list. This function does not make space to store the new node. in case the list is full, popBack() needs to be called beforehand
+			inline void pushFront(Value&& val)
+			{
+				uint32_t addr = alloc.alloc_addr(1u, 1u);
+				assert(addr < cap);
+				m_array[addr] = *(new Snode<Value>(std::move(val)));
+				auto n = m_array+addr;
+				n->prev = invalid_iterator;
+				n->next = m_begin;
+
+				if (m_begin != invalid_iterator)
+					getBegin()->prev = addr;
+				m_begin = addr;
+			}
+
+			//get node ptr of the first item in the list
+			inline Snode<Value>* getBegin() { return &(m_array[m_begin]); }
+
+			//get node ptr of the last item in the list
+			inline Snode<Value>* getBack() { return &(m_array[m_back]); }
+
+			//get index/iterator of the first element
+			inline uint32_t getFirstAddress() { return m_begin; }
+
+			//remove a node at nodeAddr from the list
+			inline void erase(uint32_t nodeAddr)
+			{
+				assert(nodeAddr != invalid_iterator);
+				auto node = get(nodeAddr);
+				if (node->prev != invalid_iterator)
+					get(node->prev)->next = node->next;
+				if (node->next != invalid_iterator)
+					get(node->next)->prev = node->prev;
+				//node->~Snode<Value>();	destructor not defined
+				alloc.free_addr(nodeAddr, 1u);
+			}
+
+			//move a node at nodeAddr to the front of the list
+			inline void moveToFront(uint32_t nodeAddr)
+			{
+				if (m_begin == nodeAddr) return;
+				getBegin()->prev = nodeAddr;
+
+				auto node = get(nodeAddr);
+				if (node->next != invalid_iterator)
+					get(node->next)->prev = node->prev;
+				if (node->prev != invalid_iterator)
+					get(node->next)->prev = node->prev;
+				node->next = m_begin;
+				node->prev = invalid_iterator;
+				m_begin = nodeAddr;
+			}
+			//Constructor, capacity determines the amount of allocated space
+			DoublyLinkedList(const uint32_t capacity) :
+				cap(capacity),
+				reservedSpace(_IRR_ALIGNED_MALLOC(PoolAddressAllocator<uint32_t>::reserved_size(1u, capacity, 1u), alignof(void*))),
+				alloc(reservedSpace, 0u, 0u, 1u, capacity, 1u)
+			{
+				m_back = invalid_iterator;
+				m_begin = invalid_iterator;
+				m_array = new Snode<Value>[capacity];
```

Comment:
i think you'll have to allocate raw storage and create `Snode`s with placement-new, just like std vector does. Otherwise you're unable to store not-default-constructible types

Reply:
@devshgraphicsprogramming ?

Reply:
this should have been a malloc/free not a new/delete